File: java/leetcode_grind/Day328.java
package leetcode_grind;

public class Day328 {
    // https://leetcode.com/problems/integer-break/description/
    static class Solution {
        public int integerBreak(int n) {
            if (n <= 3) {
                return n - 1;
            }
            var dp = new int[n + 1];
            dp[1] = 1;
            dp[2] = 2;
            dp[3] = 3;

            for (int i = 4; i <= n; i++) {
                for (int j = 1; j <= i; j++) {
                    dp[i] = Math.max(dp[i], j * dp[i - j]);
                }
            }

            return dp[n];
        }
    }
}

File: src/leetcode_grind/day_328.rs
// https://leetcode.com/problems/integer-break/description
pub fn integer_break(n: i32) -> i32 {
    let mut n = n;
    if n <= 3 {
        return n - 1;
    }

    let mut ans = 1;
    while n > 4 {
        ans *= 3;
        n -= 3;
    }

    ans * n
}
